### PR TITLE
CP-1347 Parse args leniently, forward to underlying executables

### PR DESCRIPTION
**Pull request by evanweible-wf**
## Issue
- Currently we have to explicitly support every argument that we want to pass on to the underlying executables because the `ArgParser` class throws if it encounters an unknown option.

## Changes
**Source:**
- Added a `LenientParser` class and a `LenientArgResults` class that can be used to parse args in a lenient manner that captures the unknown args instead of throwing.
- Updated most of the tasks to use this lenient arg parsing and forward all unknown args to the underlying executable (`dartanalyzer`, `dart_style:format`, `test`)

**Tests:**
- Updated tests
- Analyzer tests are still failing (will be fixed in #126)
- Coverage tests are still failing (see https://github.com/dart-lang/coverage/issues/104)

## Areas of Regression
- Any task that is forwarding args:
  - `analyze`
  - `format`
  - `test`

## Testing
- CI passes

## Code Review
@trentgrover-wf
@maxwellpeterson-wf 
@dustinlessard-wf
@jayudey-wf

### CODE REVIEW

File: lib/src/tasks/analyze/api.dart [hunk 1]
```diff
@@ -41,15 +42,21 @@ AnalyzeTask analyze(
...
   args.addAll(_findFilesFromEntryPoints(entryPoints));
 
   TaskProcess process = new TaskProcess(executable, args);
   AnalyzeTask task =
       new AnalyzeTask('$executable ${args.join(' ')}', process.done);
 
-  process.stdout.listen(task._analyzerOutput.add);
-  process.stderr.listen(task._analyzerOutput.addError);
+  process.stdout.listen((l) {
+    print(l);
```
Comment (trentgrover-wf) [inline lib/src/tasks/analyze/api.dart]: are these prints intentional?

File: lib/src/tasks/test/cli.dart [hunk 2]
```diff
@@ -34,79 +35,77 @@ class TestCli extends TaskCli {
...
-          print("Ignoring unknown argument");
-        }
-      });
-    }
-    return individualTests;
+  bool hasRestParams(LenientArgResults parsedArgs) {
+    return parsedArgs.rest.length > 0;
   }
 
+//  Future addToTestsFromRest(List<String> tests, List<String> rest) async {
```
Comment (trentgrover-wf) [inline lib/src/tasks/test/cli.dart]: delete cruft

File: lib/src/tasks/test/cli.dart [hunk 3]
```diff
@@ -34,79 +35,77 @@ class TestCli extends TaskCli {
...
+        (option) => option.contains('-p') || option.contains('--platform'))) {
+      platforms = config.test.platforms;
     }
-    List<String> platforms =
-        TaskCli.valueOf('platform', parsedArgs, config.test.platforms);
 
-    if (hasRestParams(parsedArgs)) {
-      individualTests = await addToTestsFromRest(tests, parsedArgs.rest);
-    }
+//    if (hasRestParams(parsedArgs)) {
```
Comment (trentgrover-wf) [inline lib/src/tasks/test/cli.dart]: ?

File: lib/src/tasks/test/cli.dart (hunk 2 above)
Comment (evanweible-wf) [inline lib/src/tasks/test/cli.dart]: removed

File: lib/src/tasks/test/cli.dart (hunk 3 above)
Comment (evanweible-wf) [inline lib/src/tasks/test/cli.dart]: removed, covered by arg forwarding now

File: lib/src/tasks/analyze/api.dart (hunk 1 above)
Comment (evanweible-wf) [inline lib/src/tasks/analyze/api.dart]: nope, removed

**Comment by trentgrover-wf** [comment]
+1

**Comment by dustinlessard-wf** [comment]
+1

**Comment by maxwellpeterson-wf** [comment]
+1

**Comment by trentgrover-wf** [comment]
@jayudey-wf ready for merge

**Comment by jayudey-wf** [comment]
found an issue that Evan is currently investigating

**Comment by rmconsole2-wf** [comment]
@evanweible-wf This pull request has merge conflicts, please resolve.

**Comment by trentgrover-wf** [comment]
@evanweible-wf is this stalled indefinitely?

**Comment by evanweible-wf** [comment]
@trentgrover-wf yeah.. Main motivation for it was the `-n` flag for the test task, but that's been manually added. I think I'll defer this for now, maybe revisit as a part of v2.

**Pull request closed without merging**